wandb/sdk/launch/agent/agent.py

```diff
@@ -36,6 +38,83 @@
 _logger = logging.getLogger(__name__)
 
 
+def init_pool_processes(jobs: Dict[str, int], lock: Lock) -> None:
+    global _jobs
+    global _jobs_lock
+    _jobs = jobs  # type: ignore
+    _jobs_lock = lock  # type: ignore
+
+
+def thread_run_job(
+    launch_spec: Dict[str, Any],
+    job: Dict[str, Any],
+    default_config: Dict[str, Any],
+    api: Api,
+) -> None:
+    project = create_project_from_spec(launch_spec, api)
+    _logger.info("Fetching and validating project...")
+    project = fetch_and_validate_project(project, api)
+    _logger.info("Fetching resource...")
+    resource = launch_spec.get("resource") or "local-container"
+    backend_config: Dict[str, Any] = {
+        PROJECT_SYNCHRONOUS: False,  # agent always runs async
+    }
+
+    backend_config["runQueueItemId"] = job["runQueueItemId"]
+    _logger.info("Loading backend")
+    override_build_config = launch_spec.get("build")
+    override_registry_config = launch_spec.get("registry")
+
+    build_config, registry_config = resolve_build_and_registry_config(
+        default_config, override_build_config, override_registry_config
+    )
+    builder = load_builder(build_config)
+
+    default_runner = default_config.get("runner", {}).get("type")
+    if default_runner == resource:
+        backend_config["runner"] = default_config.get("runner")
+    backend = load_backend(resource, api, backend_config)
+    backend.verify()
+    _logger.info("Backend loaded...")
+    run = backend.run(project, builder, registry_config)
+
+    if not run:
+        return
+    with _jobs_lock:  # type: ignore
+        _jobs[run.id] = 1  # type: ignore
+    try:
+        while True:
+            if _is_run_finished(run):
+                with _jobs_lock:  # type: ignore
+                    del _jobs[run.id]  # type: ignore
+                return
+            time.sleep(AGENT_POLLING_INTERVAL)
+    except KeyboardInterrupt:
```

Comment:
Does this work in a thread? Are all threads killed on a single keyboard interrrupt?

Reply:
Good callout - I found a [simple way](https://stackoverflow.com/questions/11436502/closing-all-threads-with-a-keyboard-interrupt) to have all threads exit on a KeyboardInterrupt, will add in the next commit. I don't think the solution above will actually capture the interrupt.